Mathlib/Data/Fintype/Quotient.lean

```diff
@@ -1,90 +1,241 @@
+    intro as
+    rw [List.Pi.cons_map a as (fun i ↦ Quotient.mk (S i))]
+    exact f _
+
+end List
+
+section Fintype
+variable {ι : Type*} [Fintype ι] [DecidableEq ι] {α : ι → Sort*} [S : ∀ i, Setoid (α i)] {β : Sort*}
+
+/-- Choice-free induction principle for quotients indexed by a finite type.
+  See `Quotient.induction_on_pi` for the general version assuming `Classical.choice`. -/
+@[elab_as_elim]
+lemma fin_ind {C : (∀ i, Quotient (S i)) → Prop}
+    (f : ∀ a : ∀ i, α i, C (⟦a ·⟧)) (q : ∀ i, Quotient (S i)) : C q := by
+  have {m : Multiset ι} (C : (∀ i ∈ m, Quotient (S i)) → Prop) :
+      (_ : ∀ a : ∀ i ∈ m, α i, C (⟦a · ·⟧)) (q : ∀ i ∈ m, Quotient (S i)), C q := by
+    intro m C
+    induction m using Quotient.ind
+    exact list_ind (S := S)
+  exact this (fun q ↦ C (q · (Finset.mem_univ _))) (fun _ ↦ f _) (fun i _ ↦ q i)
+
+/-- Choice-free induction principle for quotients indexed by a finite type.
+  See `Quotient.induction_on_pi` for the general version assuming `Classical.choice`. -/
+@[elab_as_elim]
+lemma fin_induction_on {C : (∀ i, Quotient (S i)) → Prop}
+    (q : ∀ i, Quotient (S i)) (f : ∀ a : ∀ i, α i, C (⟦a ·⟧)) : C q :=
+  fin_ind f q
+
+/-- Given a collection of setoids indexed by a fintype `ι` and a function that for each `i : ι`
+  gives a term of the corresponding quotient type, then there is corresponding term in the quotient
+  of the product of the setoids.
+  See `Quotient.choice` for the noncomputable general version. -/
+def finChoice (q : ∀ i, Quotient (S i)) :
+    @Quotient (∀ i, α i) piSetoid := by
+  let e := Equiv.subtypeQuotientEquivQuotientSubtype (fun l : List ι ↦ ∀ i, i ∈ l)
+    (fun s : Multiset ι ↦ ∀ i, i ∈ s) (fun i ↦ Iff.rfl) (fun _ _ ↦ Iff.rfl) ⟨_, Finset.mem_univ⟩
+  refine e.liftOn
+    (fun l ↦ (listChoice fun i _ ↦ q i).map (fun a i ↦ a i (l.2 i)) ?_) ?_
+  · exact fun _ _ h i ↦ h i _
+  intro _ _ _
+  refine fin_ind (fun a ↦ ?_) q
+  simp_rw [listChoice_mk, Quotient.map_mk]
+
+theorem finChoice_eq (a : ∀ i, α i) :
+    finChoice (⟦a ·⟧) = ⟦a⟧ := by
+  dsimp [finChoice]
+  obtain ⟨l, hl⟩ := (Finset.univ.val : Multiset ι).exists_rep
+  simp_rw [← hl, Equiv.subtypeQuotientEquivQuotientSubtype, listChoice_mk]
+  rfl
+
+lemma eval_finChoice (f : ∀ i, Quotient (S i)) :
+    eval (finChoice f) = f :=
+  fin_induction_on f (fun a ↦ by rw [finChoice_eq]; rfl)
+
+/-- Lift a function on `∀ i, α i` to a function on `∀ i, Quotient (S i)`. -/
+def finLiftOn (q : ∀ i, Quotient (S i)) (f : (∀ i, α i) → β)
+    (h : ∀ (a b : ∀ i, α i), (∀ i, a i ≈ b i) → f a = f b) : β :=
+  (finChoice q).liftOn f h
+
+@[simp]
+lemma finLiftOn_empty [e : IsEmpty ι] (q : ∀ i, Quotient (S i)) :
+    @finLiftOn _ _ _ _ _ β q = fun f _ ↦ f e.elim := by
+  ext f h
+  dsimp [finLiftOn]
+  induction finChoice q using Quotient.ind
+  exact h _ _ e.elim
+
+@[simp]
+lemma finLiftOn_mk (a : ∀ i, α i) :
+    @finLiftOn _ _ _ _ _ β (⟦a ·⟧) = fun f _ ↦ f a := by
+  ext f h
+  dsimp [finLiftOn]
+  rw [finChoice_eq]
   rfl
```

Comment:
It's not immediately clear that `finLiftOn` is useful, given you can spell it `(finChoice q).liftOn f h`. Can you explain the motivation for introducing it in the PR description?